tests/unit/s2n_self_talk_client_hello_cb_test.c

```diff
@@ -181,17 +182,36 @@ int client_hello_fail_handshake(struct s2n_connection *conn, void *ctx)
     /* Return negative value to terminate the handshake */
     return -1;
 
-} 
+}
+
+int s2n_client_hello_3_poll_cb(struct s2n_connection *conn, void *ctx)
+{
+    struct client_hello_context *client_hello_ctx;
+    if (ctx == NULL) {
+        return -1;
+    }
+    client_hello_ctx = ctx;
+    /* Increment counter to ensure that callback was invoked */
+    client_hello_ctx->invoked++;
 
-int s2n_negotiate_nonblocking_ch_cb(struct s2n_connection *conn, 
+    /* complete on the 3rd call */
+    if (client_hello_ctx->invoked == 3) {
```

Comment:
This seems a little brittle. Maybe pass a pointer to "bool is_done" instead of the current context? It doesn't look like you use anything else on the client_hello_context struct other than invoked.

Reply:
A bool is not sufficient imo to capture the "function got called more than one time". We could add a "require_poll_count" to the context which would capture the `3` and be passed in from the test call site.

However the magic `3` is also part of the callback function name so there is some self documentation already.

Reply:
I'd rather you didn't have a "required_poll_count" and instead had some independent way to communicate to your callback that it should complete. Triggering completion based on the number of times invoked seems potentially dangerous, since the number of times invoked is part of what you're testing. You specifically don't want your callback to be affected by the number of times it gets polled ;)